File: server/src/services/prompt/promptService.js
// server/src/services/prompt/promptService.js
// Prompt 调试：模板管理、A/B 测试评分
import { z } from 'zod'
import { createChatModel } from '../model.js'

// 评分模型用 temperature=0，结果稳定
const scoreModel = createChatModel({ temperature: 0 })

// ── Prompt 模板存储（生产用数据库）────────────────────────────
const templateStore = new Map()
let templateIdSeq = 1

// 内置示例模板
const defaultTemplates = [
  {
    id: 't_default_1',
    name: '前端助手',
    systemPrompt: '你是前端开发专家，精通 Vue3、React、TypeScript。回答简洁准确，必要时给代码示例。',
    description: '通用前端技术问答',
    tags: ['前端', '技术'],
    createdAt: new Date().toISOString(),
    versions: [],
  },
  {
    id: 't_default_2',
    name: '代码 Review',
    systemPrompt: `你是资深代码评审专家。审查代码时，按以下顺序输出：
1. 【总体评价】一句话概括
2. 【问题列表】按严重程度排序，每条格式：[严重/一般/建议] 具体问题
3. 【优化建议】具体的改进代码示例
语气专业，直指问题，不废话。`,
    description: '代码审查专用',
    tags: ['代码', '审查'],
    createdAt: new Date().toISOString(),
    versions: [],
  },
  {
    id: 't_default_3',
    name: '简洁问答',
    systemPrompt: '用最简洁的语言回答问题，不超过3句话，不用废话开场。',
    description: '简短精准的回答风格',
    tags: ['简洁'],
    createdAt: new Date().toISOString(),
    versions: [],
  },
]

defaultTemplates.forEach(t => templateStore.set(t.id, t))

// ── 模板 CRUD ─────────────────────────────────────────────────
export function listTemplates() {
  return [...templateStore.values()].sort((a, b) =>
    new Date(b.createdAt) - new Date(a.createdAt)
  )
}

export function getTemplate(id) {
  return templateStore.get(id) || null
}

export function saveTemplate({ name, systemPrompt, description = '', tags = [], existingId }) {
  const id = existingId || `t_${Date.now()}_${templateIdSeq++}`
  const existing = templateStore.get(id)

  const template = {
    id,
    name,
    systemPrompt,
    description,
    tags,
    createdAt: existing?.createdAt || new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    // 保存版本历史（最多保留 10 个版本）
    versions: [
      ...(existing?.versions || []),
      {
        version:      (existing?.versions?.length || 0) + 1,
        systemPrompt: existing?.systemPrompt || systemPrompt,
        savedAt:      new Date().toISOString(),
      },
    ].slice(-10),
  }

  templateStore.set(id, template)
  return template
}

export function deleteTemplate(id) {
  // 默认模板不允许删除
  if (id.startsWith('t_default_')) throw new Error('内置模板不能删除')
  const existed = templateStore.delete(id)
  if (!existed) throw new Error('模板不存在')
}

// ── A/B 测试：AI 自动评分 ──────────────────────────────────────
const EvalSchema = z.object({
  relevance:   z.number().min(1).max(5).describe('回答与问题的相关性'),
  accuracy:    z.number().min(1).max(5).describe('内容的准确性'),
  clarity:     z.number().min(1).max(5).describe('表达的清晰度'),
  conciseness: z.number().min(1).max(5).describe('是否简洁，不啰嗦'),
  overall:     z.number().min(1).max(5).describe('综合评分'),
  winner:      z.enum(['A', 'B', 'tie']).describe('哪个回答更好，相差不大就选 tie'),
  reason:      z.string().describe('判断理由，一句话，不超过40字'),
})

export async function scoreAbTest({ question, answerA, answerB }) {
  const evalModel = scoreModel.withStructuredOutput(EvalSchema)

  // 同时对 A、B 各自评分
  const [evalA, evalB] = await Promise.all([
    evalModel.invoke([
      { role: 'system', content: '你是 AI 回答质量评估专家，客观评分，不偏袒任何一方。' },
      { role: 'user',   content: `问题：${question}\n\n回答：${answerA}` },
    ]),
    evalModel.invoke([
      { role: 'system', content: '你是 AI 回答质量评估专家，客观评分，不偏袒任何一方。' },
      { role: 'user',   content: `问题：${question}\n\n回答：${answerB}` },
    ]),
  ])

  // 综合对比，判断胜者
  const compareModel = scoreModel.withStructuredOutput(z.object({
    winner: z.enum(['A', 'B', 'tie']),
    reason: z.string().describe('对比理由，30字以内'),
  }))

  const comparison = await compareModel.invoke([
    { role: 'system', content: '比较两个回答，选出更好的那个。评分相差0.5分以内视为平局。' },
    { role: 'user',   content: `
问题：${question}

回答A：${answerA}
A的评分：相关性${evalA.relevance} 准确性${evalA.accuracy} 清晰度${evalA.clarity} 简洁性${evalA.conciseness} 综合${evalA.overall}

回答B：${answerB}
B的评分：相关性${evalB.relevance} 准确性${evalB.accuracy} 清晰度${evalB.clarity} 简洁性${evalB.conciseness} 综合${evalB.overall}

哪个回答更好？` },
  ])

  return {
    scoreA: evalA,
    scoreB: evalB,
    winner: comparison.winner,
    reason: comparison.reason,
  }
}

File: server-py/app/services/prompt/prompt_service.py
# server-py/app/services/prompt/prompt_service.py
# Prompt 调试：模板管理、A/B 测试评分
import asyncio
import time
from datetime import datetime, timezone
from typing import Literal

from pydantic import BaseModel, Field

from app.services.model import create_chat_model

# 评分模型用 temperature=0，结果稳定
_score_model = create_chat_model(temperature=0)

# ── Prompt 模板存储（生产用数据库）────────────────────────────
_template_store: dict[str, dict] = {}
_template_id_seq = 1


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


# 内置示例模板
_default_templates = [
    {
        "id": "t_default_1",
        "name": "前端助手",
        "systemPrompt": "你是前端开发专家，精通 Vue3、React、TypeScript。回答简洁准确，必要时给代码示例。",
        "description": "通用前端技术问答",
        "tags": ["前端", "技术"],
        "createdAt": _now_iso(),
        "versions": [],
    },
    {
        "id": "t_default_2",
        "name": "代码 Review",
        "systemPrompt": """你是资深代码评审专家。审查代码时，按以下顺序输出：
1. 【总体评价】一句话概括
2. 【问题列表】按严重程度排序，每条格式：[严重/一般/建议] 具体问题
3. 【优化建议】具体的改进代码示例
语气专业，直指问题，不废话。""",
        "description": "代码审查专用",
        "tags": ["代码", "审查"],
        "createdAt": _now_iso(),
        "versions": [],
    },
    {
        "id": "t_default_3",
        "name": "简洁问答",
        "systemPrompt": "用最简洁的语言回答问题，不超过3句话，不用废话开场。",
        "description": "简短精准的回答风格",
        "tags": ["简洁"],
        "createdAt": _now_iso(),
        "versions": [],
    },
]

for _t in _default_templates:
    _template_store[_t["id"]] = _t


# ── 模板 CRUD ─────────────────────────────────────────────────
def list_templates() -> list[dict]:
    return sorted(_template_store.values(), key=lambda t: t["createdAt"], reverse=True)


def get_template(template_id: str) -> dict | None:
    return _template_store.get(template_id)


def save_template(name: str, system_prompt: str, description: str = "", tags: list[str] | None = None,
                   existing_id: str | None = None) -> dict:
    global _template_id_seq
    tags = tags or []
    template_id = existing_id or f"t_{int(time.time() * 1000)}_{_template_id_seq}"
    _template_id_seq += 1
    existing = _template_store.get(template_id)

    versions = [*(existing.get("versions", []) if existing else []), {
        "version": len(existing.get("versions", [])) + 1 if existing else 1,
        "systemPrompt": existing["systemPrompt"] if existing else system_prompt,
        "savedAt": _now_iso(),
    }][-10:]

    template = {
        "id": template_id,
        "name": name,
        "systemPrompt": system_prompt,
        "description": description,
        "tags": tags,
        "createdAt": existing["createdAt"] if existing else _now_iso(),
        "updatedAt": _now_iso(),
        "versions": versions,
    }

    _template_store[template_id] = template
    return template


def delete_template(template_id: str):
    if template_id.startswith("t_default_"):
        raise ValueError("内置模板不能删除")
    if template_id not in _template_store:
        raise ValueError("模板不存在")
    del _template_store[template_id]


# ── A/B 测试：AI 自动评分 ──────────────────────────────────────
class _Evaluation(BaseModel):
    relevance: int = Field(ge=1, le=5, description="回答与问题的相关性")
    accuracy: int = Field(ge=1, le=5, description="内容的准确性")
    clarity: int = Field(ge=1, le=5, description="表达的清晰度")
    conciseness: int = Field(ge=1, le=5, description="是否简洁，不啰嗦")
    overall: int = Field(ge=1, le=5, description="综合评分")


class _Comparison(BaseModel):
    winner: Literal["A", "B", "tie"]
    reason: str = Field(description="对比理由，30字以内")


async def score_ab_test(question: str, answer_a: str, answer_b: str) -> dict:
    eval_model = _score_model.with_structured_output(_Evaluation)

    eval_a, eval_b = await asyncio.gather(
        eval_model.ainvoke([
            {"role": "system", "content": "你是 AI 回答质量评估专家，客观评分，不偏袒任何一方。"},
            {"role": "user", "content": f"问题：{question}\n\n回答：{answer_a}"},
        ]),
        eval_model.ainvoke([
            {"role": "system", "content": "你是 AI 回答质量评估专家，客观评分，不偏袒任何一方。"},
            {"role": "user", "content": f"问题：{question}\n\n回答：{answer_b}"},
        ]),
    )

    compare_model = _score_model.with_structured_output(_Comparison)
    comparison: _Comparison = await compare_model.ainvoke([
        {"role": "system", "content": "比较两个回答，选出更好的那个。评分相差0.5分以内视为平局。"},
        {"role": "user", "content": f"""
问题：{question}

回答A：{answer_a}
A的评分：相关性{eval_a.relevance} 准确性{eval_a.accuracy} 清晰度{eval_a.clarity} 简洁性{eval_a.conciseness} 综合{eval_a.overall}

回答B：{answer_b}
B的评分：相关性{eval_b.relevance} 准确性{eval_b.accuracy} 清晰度{eval_b.clarity} 简洁性{eval_b.conciseness} 综合{eval_b.overall}

哪个回答更好？"""},
    ])

    return {
        "scoreA": eval_a.model_dump(),
        "scoreB": eval_b.model_dump(),
        "winner": comparison.winner,
        "reason": comparison.reason,
    }
